React storage/cancel upload from process file

raystorm opened the pull request:
#### Description of changes
Adds support for the ability to cancel an Upload from `processFile()` 
by returning a rejected promise.

#### Issue #, if available
#5099 

#### Description of how you validated changes
  + Added a Unit Test to ensure `removeUpload()` is called when a rejected promise is returned from `processFile()`.
  + Added a New Example, that forces a `Promise.reject()` in examples. 
    ran the example manually with additional Logging code
    (not committed, and now removed to watch/verify operation.)

#### Checklist

- [x] Have read the [Pull Request Guidelines](https://github.com/aws-amplify/amplify-ui/blob/main/CONTRIBUTING.md)
- [x] PR description included
- [ ] Relevant documentation is changed or added (and PR referenced)
- [x] `yarn test` passes and tests are updated/added
- [x] No side effects or [`sideEffects`](https://github.com/aws-amplify/amplify-ui/blob/main/packages/react/CONTRIBUTING.md#code-standards) field updated

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

calebpollman commented:
@raystorm Thanks for opening this! Ned to discuss with the team, will update here once we have next steps

calebpollman commented:
@raystorm Just wanted to update you here, we are aligned with adding the update to `processFile` and will be reviewing your PR. Thanks again for contributing!

calebpollman commented:
@raystorm We recently made some updates to the `StorageManager` to support new functionality exposed on the `uploadData` API (consumed from `aws-amplify/storage`) used by the component, which has caused some merge conflicts with between `main` and your PR. Can help out with resolvng the merge conflicts, but wanted to give you the option to handle resolvng them i you prefer

raystorm commented:
> @raystorm We recently made some updates to the `StorageManager` to support new functionality exposed on the `uploadData` API (consumed from `aws-amplify/storage`) used by the component, which has caused some merge conflicts with between `main` and your PR. Can help out with resolvng the merge conflicts, but wanted to give you the option to handle resolvng them i you prefer

Merge Conflicts are resolved.  You may want to run the test example from my branch locally, to make sure everything still works properly.  The Error Handling code is the same, but moved to `getInput()` I lost my local setup for running the examples, when I cleaned up some accidentally pushed env changes on my side.  So I was unable to re-run the examples locally. as a secondary validation of my changes this time.

raystorm commented:
Not to be impatient, but is there any expected timeline for review/merge/release?

calebpollman commented:
Hey @raystorm, valid question. Have started reviewing, will finish the review in the next couple days. Thanks for your patience!

raystorm commented on CONTRIBUTING.md:
```diff
@@ -43,7 +43,7 @@ You should open an issue to discuss your pull request, unless it's a trivial cha
 1. [`nvm install`](https://github.com/nvm-sh/nvm)
 1. [`nvm use`](https://github.com/nvm-sh/nvm)
-1. `yarn setup`
+1. `yarn setup` - _NOTE:_ if Cypress Fails to download during this step, delete `yarn.lock` and try again.
```
I will remove the note.  
But for the record, I disagree.
If deleting that file has a large impact, maybe more should be specified in the package.json files.
Also how would you recommend to fix the Cypress issue?

raystorm commented on packages/react/CONTRIBUTING.md:
```diff
@@ -3,20 +3,29 @@
 ## Dependencies
 
-`@aws-amplify/ui-react` depends on [`@aws-amplify/ui`](../ui) for theming, state management, and translation logic. If you're looking for change in these, please refer to `@aws-amplify/ui` [README](../ui/README.md).
+`@aws-amplify/ui-react` depends on [`@aws-amplify/ui`](../ui) for theming,
+state management, and translation logic.
+If you're looking for change in these,
+please refer to `@aws-amplify/ui` [README](../ui/README.md).
```
I really dislike scrolling sideways and that line is really long. They way I broke it up, (without any whitespace at the end of the line) still renders as a single paragraph.

reesscot replied:
Scrolling sideways can be fixed by setting your code editor to wrap lines at a certain length

raystorm replied:
> Scrolling sideways can be fixed by setting your code editor to wrap lines at a certain length

Yes, but that will wrap, wherever. By Manually wrapping I can more naturally flow the line.

raystorm commented on packages/react/CONTRIBUTING.md:
```diff
@@ -3,20 +3,29 @@
 1. Run `yarn react dev`
 
-This will start building `@aws-amplify/ui-react` in watch mode. To test your changes, you can utilize [`examples/next`](../../examples/next) to run examples on `next.js`. Please see examples [README](../../examples/README.md) and e2e [README](../e2e/README.md#contributing) to get started.
+This will start building `@aws-amplify/ui-react` in watch mode.
+To test your changes, you can utilize [`examples/next`](../../examples/next)
+to run examples on `next.js`.
+Please see examples [README](../../examples/README.md) and e2e [README](../e2e/README.md#contributing) to get started.
```
I really dislike scrolling sideways and that line is really long. They way I broke it up, (without any whitespace at the end of the line) still renders as a single paragraph.

raystorm commented on turbo.json:
```diff
@@ -1,6 +1,6 @@
 {
   "$schema": "https://turborepo.org/schema.json",
-  "baseBranch": "origin/main",
+  // "baseBranch": "origin/main",
```
Turbo didn't work for me with that line uncommented.

raystorm commented on packages/react-storage/src/components/StorageManager/utils/getInput.ts:
```diff
@@ -35,7 +42,17 @@ export const getInput = ({
+          onUploadError(error ?? `Error processing: ${key}`, { key: key });
+        }
+        removeUpload({ id });
+        return result;
```
Line 52, removes the Upload.
I was returning the rejected promise, after cancelling the upload in an attempt to be minimally disruptive.

raystorm commented on packages/react-storage/src/components/StorageManager/hooks/useUploadFiles/__tests__/useUploadFiles.spec.ts:
```diff
@@ -190,6 +193,32 @@ describe('useUploadFiles', () => {
     });
   });
 
+  it('should remove upload after processFile promise rejects', async () => {
+    const processFile: StorageManagerProps['processFile'] = ({ file }) =>
+      Promise.reject({ file, key: 'test.png', error: 'forced test error' });
+
+    const { waitForNextUpdate } = renderHook(() =>
+      useUploadFiles({
+        ...props,
+        isResumable: true,
+        processFile,
+        files: [mockQueuedFile],
+      })
+    );
+
+    waitForNextUpdate();
+
+    await waitFor(() => {
+      expect(mockOnUploadError).toHaveBeenCalledWith('forced test error', {
+        key: 'test.png',
+      });
+    });
+
+    await waitFor(() => {
+      expect(mockRemoveUpload).toHaveBeenCalled();
+    });
+  });
```
@calebpollman,  With your comments on my example is this Unit Test incorrect?
Should I only be testing a thrown Exception? or is that a *second* test case?

raystorm commented on packages/react-storage/src/components/StorageManager/utils/getInput.ts:
```diff
@@ -35,7 +42,17 @@ export const getInput = ({
+    } = await resolveFile({ file, key, processFile }).catch(
+      (result: ProcessFileParams) => {
+        const { key, error } = result;
+        if (isFunction(onUploadError)) {
```
This has been commented out as incorrect, with a note, for a handler that might need to be added.  Let me know if I should remove the code and the comment.

raystorm commented:
@calebpollma  I've made a lot of the changes you requested as I understood them.  
It appears we have different visions for how this is to behave.  I was just running the example locally, and for me, when I run `examples/next/pages/ui/components/storage/storage-manager/process-file-reject` it silently fails.  So how should the end-user be notified that the uploaded file was rejected?

raystorm commented:
@calebpollman Looking at the latest merge conflicts with the new `onProcessFileSuccess()` method, are you now handling this ask internally in a different way? Should I expect this PR to be closed and not merged in favor of your internal solution?

raystorm commented:
> @calebpollman I've made a lot of the changes you requested as I understood them. It appears we have different visions for how this is to behave. I was just running the example locally, and for me, when I run `examples/next/pages/ui/components/storage/storage-manager/process-file-reject` it silently fails. So how should the end-user be notified that the uploaded file was rejected?

@reesscot Are you on the team for this as well? 

Having not seen a response, I am unsure how to proceed?
Thinking about it, I think a new optional function is needed, for the end user.  `onProcessFileError()` or similar.  Will there be any problems if I add a new function to the API?  Or are you solving this internally?

calebpollman commented:
Hey @raystorm, there's some problem areas in the upload input resolution we need to sort out in order to correctly address this. Would still like to get this in, but thinking that it might be more straightforward if we integrate your changes in with the input resolution refactoring, does that wor for you?

raystorm commented:
I just want this resolved. 
This is a blocking issue in my application right now. 
I didn't expect this to take this long. 
I'm just confused on what's going on, and the nature if the most recent updates with the onSuccess method, and what that means for my change. 

Is there something I need to do to help move things forward?

calebpollman commented:
@raystorm Apologies here - there has been some churn on our end and we will work with you to get this resolved. 

Regarding `onProcessFileSuccess`, it was added to address an issue with the value provided to `onUploadSuccess` not being the value that was provided to the upload call if updated through `processFile`, it was not intended nor it should impact the behavior you are implementing here.

raystorm commented:
> @raystorm Apologies here - there has been some churn on our end and we will work with you to get this resolved.
> 
> Regarding `onProcessFileSuccess`, it was added to address an issue with the value provided to `onUploadSuccess` not being the value that was provided to the upload call if updated through `processFile`, it was not intended nor it should impact the behavior you are implementing here.

I've done my best to FINALLY, get the merge conflicts resolved AGAIN.
I've also added a new function `onProcessFileError` and a new type, `onProcessFileErrorParams`.  I've moved the error handling into `resolveFile` where it should have been in the first place so it directly responds to process file throwing an exception or rejecting.  
Hopefully the new code meets with approval and we can finally get this PR done.

reesscot commented on .changeset/short-doors-flash.md:
```diff
@@ -0,0 +1,5 @@
+"@aws-amplify/ui-react-storage": patch
+---
+
+fix(storage-manager): update file processedKey on process file complete
```
I don't think this file is needed

reesscot commented on docs/src/pages/[platform]/connected-components/liveness/customization/customization.system-clock-offset.react.mdx:
```diff
@@ -14,30 +14,25 @@ function MyComponent() {
   /* 
-   * 1. Check whether the difference between server time and device time is
-   *  greater than or equal to 5 minutes, and if so, return the offset in milliseconds.
-   *  This logic should be adjusted based on the server response and use case
```
Are these changes related to your PR? Same comment on all the other liveness files.

reesscot commented on packages/react-core/src/hooks/__tests__/useDataState.spec.ts:
```diff
@@ -26,7 +26,6 @@ describe('useDataState', () => {
       expect(action).not.toHaveBeenCalled();
 
       expect(initState.data).toBe(initData);
-      expect(initState.hasError).toBe(false);
```
This file shouldn't be getting changed

reesscot reviewed (changes requested):
There's a lot of files being changed in this PR that aren't related to your change, so we're not going to be able to merge this as is. Can you please squash your changes and rebase them against the latest version of main? Looks like something went wrong with resolving merge conflicts.

raystorm commented:
> There's a lot of files being changed in this PR that aren't related to your change, so we're not going to be able to merge this as is. Can you please squash your changes and rebase them against the latest version of main? Looks like something went wrong with resolving merge conflicts.

@reesscot These changes are a result of a merge issue when pulling in the latest version from main to fix some merge issues, blocking the PR from being mergable.  I'm not good enough with Git to quickly squash and rebase.  So instead I've created a new branch and PR #5717 
I am closing this PR for that one.
